internal/services/mssql/mssql_virtual_machine_availability_group_listener_resource.go

```diff
@@ -0,0 +1,510 @@
+
+	ipAddress := ""
+	if input.IPAddress != nil {
+		ipAddress = *input.IPAddress
+	}
+	privateIpAddress.IpAddress = ipAddress
+
+	subnetId := ""
+	if input.SubnetResourceId != nil {
+		id, err := networkParse.SubnetID(*input.SubnetResourceId)
+		if err != nil {
+			return nil, err
+		}
+		subnetId = id.ID()
+	}
+
+	privateIpAddress.SubnetId = subnetId
+
+	return []helper.PrivateIpAddressMsSqlVirtualMachineAvailabilityGroupListener{privateIpAddress}, nil
+}
+
+func expandMsSqlVirtualMachineAvailabilityGroupListenerReplicas(replicas []helper.ReplicaMsSqlVirtualMachineAvailabilityGroupListener) (*[]availabilitygrouplisteners.AgReplica, error) {
+	results := make([]availabilitygrouplisteners.AgReplica, 0)
+
+	for _, rep := range replicas {
+		replica := availabilitygrouplisteners.AgReplica{
+			Role:              pointer.To(availabilitygrouplisteners.Role(rep.Role)),
+			Commit:            pointer.To(availabilitygrouplisteners.Commit(rep.Commit)),
+			Failover:          pointer.To(availabilitygrouplisteners.Failover(rep.Failover)),
+			ReadableSecondary: pointer.To(availabilitygrouplisteners.ReadableSecondary(rep.ReadableSecondary)),
+		}
+
+		sqlVirtualMachineId := rep.SqlVirtualMachineId
+		if sqlVirtualMachineId != "" {
+			id, err := sqlvirtualmachines.ParseSqlVirtualMachineID(sqlVirtualMachineId)
+			if err != nil {
+				return nil, err
+			}
+			sqlVirtualMachineId = id.ID()
+		}
+		replica.SqlVirtualMachineInstanceId = pointer.To(sqlVirtualMachineId)
+
+		results = append(results, replica)
+	}
+	return &results, nil
+}
+
+func flattenMsSqlVirtualMachineAvailabilityGroupListenerReplicas(input *[]availabilitygrouplisteners.AgReplica, subscriptionId string) ([]helper.ReplicaMsSqlVirtualMachineAvailabilityGroupListener, error) {
+	results := make([]helper.ReplicaMsSqlVirtualMachineAvailabilityGroupListener, 0)
+	if input == nil || len(*input) == 0 {
+		return results, nil
+	}
+
+	for _, replica := range *input {
+
+		sqlVirtualMachineInstanceId := ""
+		if replica.SqlVirtualMachineInstanceId != nil {
+			parsedId, err := sqlvirtualmachines.ParseSqlVirtualMachineIDInsensitively(*replica.SqlVirtualMachineInstanceId)
+			if err != nil {
+				return nil, err
+			}
+
+			newId := sqlvirtualmachines.NewSqlVirtualMachineID(subscriptionId, parsedId.ResourceGroupName, parsedId.SqlVirtualMachineName)
+			if err != nil {
+				return nil, err
+			}
+
+			sqlVirtualMachineInstanceId = newId.ID()
+		}
+
+		v := helper.ReplicaMsSqlVirtualMachineAvailabilityGroupListener{
+			SqlVirtualMachineId: sqlVirtualMachineInstanceId,
+		}
+
+		if replica.Role != nil {
+			v.Role = string(pointer.From(replica.Role))
+		}
+
+		if replica.Commit != nil {
+			v.Commit = string(pointer.From(replica.Commit))
+		}
+
+		if replica.Failover != nil {
+			v.Failover = string(pointer.From(replica.Failover))
+		}
+
+		if replica.ReadableSecondary != nil {
+			v.ReadableSecondary = string(pointer.From(replica.ReadableSecondary))
+		}
```

Comment:
These should always have a value set in state
```suggestion
		v.Role = string(pointer.From(replica.Role))
		v.Commit = string(pointer.From(replica.Commit))
		v.Failover = string(pointer.From(replica.Failover))
		v.ReadableSecondary = string(pointer.From(replica.ReadableSecondary))
```